### ISSUE TITLE
Improve release process

### ISSUE THREAD

**Issue by pdehaan** [mozilla/universal-search-addon#46]
Re: https://github.com/mozilla/universal-search-addon/issues/44#issuecomment-120551539 and https://github.com/mozilla/universal-search-addon#release-process,

Currently it seems like if we create a release, it has the potential to get pretty bloated since it can include the node_modules/*\* directory and other unwanted stuff.

It may be preferable to include all the stuff that we want bundled into the XPI into a /src/ directory or something, so we can exclude certain files/folders (like the node_modules/, tests/, README.md, package.json, config files, and other things not directly related/required for the add-on.

**Comment by pdehaan** [comment]
If we want to go w/ Grunt, we could also use stuff like:
- [**grunt-contrib-clean**](https://github.com/gruntjs/grunt-contrib-clean) instead of `rm -rf` in the package.json.
- [**grunt-contrib-compress**](https://github.com/gruntjs/grunt-contrib-compress) instead of `zip -r ...` in the package.json.

That would let us be marginally more OS-dependent.

There is also a [**grunt-scp**](https://www.npmjs.com/package/grunt-scp) module, but maybe that's getting too meta.
Although it'd be semi-convenient if we could do something like `$ npm run deploy` and have it bump the package.json version, regenerate the {install,update}.RDF files with updated version numbers, create the XPI and then deploy the necessary files to a remote server.

---

**UPDATE:** See https://github.com/mozilla/universal-search-addon/issues/46#issuecomment-122136154 for sample Gruntfile.js.

**Comment by pdehaan** [comment]
@6a68 Which files do we need for the *.xpi?
- [ ] /chrome/*
- [ ] /lib/*
- [ ] /bootstrap.js
- [ ] /chrome.manifest
- [ ] /install.rdf
- [ ] /update.rdf

**Comment by pdehaan** [comment]
Not sure if I'm happy with this, but it "works":

``` sh
➜  universal-search-addon git:(master) ✗ tree
.
├── README.md
├── dist/
│   └── addon.xpi
├── package.json
├── src/
│   ├── bootstrap.js
│   ├── chrome/
│   │   ├── content/
│   │   │   └── content.xml
│   │   └── skin/
│   │       └── binding.css
│   ├── chrome.manifest
│   ├── install.rdf
│   ├── lib/
│   │   ├── main.js
│   │   └── ui/
│   │       ├── GoButton.js
│   │       ├── Popup.js
│   │       └── Urlbar.js
│   └── update.rdf
└── test/
    └── test-index.js

9 directories, 14 files
➜  universal-search-addon git:(master) ✗ pushd src; zip -r ../dist/addon.xpi *; popd;

...

➜  universal-search-addon git:(master) ✗ ls -lh dist/*.xpi
-rw-r--r--  1 pdehaan  staff    12K Jul 16 16:37 dist/addon.xpi # only 12KB!
```

Basically I moved the following files into `src/` folder and then I just zip up the `src/*` folder and move the addon.xpi file into `dist/`.
It seems this isn't so easy on Mac to ZIP up a folder but not include that root folder, so I had to do some `pushd` and `popd` hackery to change into the `src/**` directory. Otherwise, my XPI puts all the files in a lame /src/ subdirectory.

But at least this means our XPI won't include the README.md, package.json, dist/, or test/ files+folders.

```
➜  universal-search-addon git:(master) ✗ unzip dist/addon.xpi
Archive:  dist/addon.xpi
  inflating: bootstrap.js
   creating: chrome/
   creating: chrome/content/
  inflating: chrome/content/content.xml
   creating: chrome/skin/
  inflating: chrome/skin/binding.css
  inflating: chrome.manifest
  inflating: install.rdf
   creating: lib/
  inflating: lib/main.js
   creating: lib/ui/
  inflating: lib/ui/GoButton.js
  inflating: lib/ui/Popup.js
  inflating: lib/ui/Urlbar.js
  inflating: update.rdf
```

**Comment by pdehaan** [comment]
Or, if we fancy Grunt, this works pretty well:

``` js
module.exports = function (grunt) {
  grunt.loadNpmTasks('grunt-contrib-compress');

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    compress: {
      src: {
        options: {
          archive: 'dist/addon.xpi',
          mode: 'zip'
        },
        files: [
          {expand: true, cwd: 'src/', src: ['**']}
        ]
      }
    }
  });
};
```

Or the more verbose:

``` js
module.exports = function (grunt) {
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-compress');
  grunt.loadNpmTasks('grunt-contrib-copy');

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    Constants: {
      'SRC': 'src',
      'DIST': 'dist'
    },

    clean: {
      dist: [
        '<%= Constants.DIST %>'
      ]
    },

    compress: {
      src: {
        options: {
          archive: '<%= Constants.DIST %>/addon.xpi',
          mode: 'zip'
        },
        files: [
          {expand: true, cwd: '<%= Constants.SRC %>', src: ['**']}
        ]
      }
    },

    copy: {
      rdf: {
        files: [
          {src: '<%= Constants.SRC %>/*.rdf', dest: '<%= Constants.DIST %>', expand: true, flatten: true}
        ]
      }
    }
  });

  grunt.registerTask('release', ['clean:dist', 'compress:src', 'copy:rdf']);
};
```

If we fancy Grunt, we could easily chain the `grunt eslint` task at the very beginning, as well as some grunt script to generate the RDF files from a template (see #44), and then somehow hook the `npm version` code in there somewhere. In [**fxa-content-server**](https://www.npmjs.com/package/grunt-bump) land, we use [**grunt-bump**](https://github.com/mozilla/fxa-content-server/blob/master/grunttasks/bump.js) to handle our versioning as I think it handles the GitHub tags and other stuff.

**Comment by jaredhirsch** [comment]
closed by #55
